src/shared/components/base-authoring.tsx

```diff
@@ -1,40 +1,56 @@
-import React from "react";
-import Form, { Field, IChangeEvent, UiSchema } from "react-jsonschema-form";
+import React, { useEffect, useRef } from "react";
+import Form, { Field, FormValidation, IChangeEvent, UiSchema } from "react-jsonschema-form";
 import { JSONSchema6 } from "json-schema";
+import { useDelayedValidation } from "../hooks/use-delayed-validation";
 
 import "../../shared/styles/boostrap-3.3.7.css"; // necessary to style react-jsonschema-form
 import css from "../../shared/styles/authoring.scss";
 
 export interface IBaseAuthoringProps<IAuthoredState> {
   authoredState: IAuthoredState;
-  setAuthoredState?: (state: IAuthoredState) => void;
+  setAuthoredState: (state: IAuthoredState) => void;
   preprocessFormData?: (data: IAuthoredState) => IAuthoredState;
   // react-jsonschema-form properties:
   schema: JSONSchema6;
-  uiSchema: UiSchema;
+  uiSchema?: UiSchema;
   // react-jsonschema-form additional fields.
   fields?: { [name: string]: Field };
+  validate?: (formData: IAuthoredState, errors: FormValidation) => FormValidation;
 }
 
-export const BaseAuthoring = <IAuthoredState,>({ authoredState, setAuthoredState, preprocessFormData, schema, uiSchema, fields }: IBaseAuthoringProps<IAuthoredState>) => {
+export const BaseAuthoring = <IAuthoredState,>({ authoredState, setAuthoredState, preprocessFormData, schema, uiSchema, fields, validate }: IBaseAuthoringProps<IAuthoredState>) => {
+  const formRef = useRef<Form<IAuthoredState>>(null);
+  const triggerDelayedValidation = useDelayedValidation({ formRef });
+
   const onChange = (event: IChangeEvent) => {
     let formData = event.formData;
     if (preprocessFormData) {
       formData = preprocessFormData(formData);
     }
     // Immediately save the data.
     setAuthoredState?.(formData);
+    if (validate) {
+      triggerDelayedValidation();
```

Comment:
Could also use (here and below):
```suggestion
      validate && triggerDelayedValidation();
```

Reply:
I applied other suggestions, thanks! I might skip this one for now though, as I think in most of the code I use explicit `if` statements in similar cases (except for JSX where it's hard). So, just to be consistent. Personally, I like this type of syntax, but I stopped using it over time as in some cases people mention that it's not super readable. It is to me, but I understand the other point too. ;)